### PR TITLE
RSDK-4915: implement PID control

### PR DESCRIPTION
**Pull request by martha-johnston**
This PR adds use of the control package into motor_encoder. There is still one bug in which GoTo position 0 doesn't work, but since there's lots of changes I wanted to open this now while I continue to debug that issue.

Edit: GoTo bug is fixed

Edit pt2: also should resolve [RSDK-5508](https://viam.atlassian.net/browse/RSDK-5508)

Edit pt3: I actually did not fix this above issue^ can't seem to reproduce at all

[RSDK-5508]: https://viam.atlassian.net/browse/RSDK-5508?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: components/motor/gpio/motor_encoder.go
Comment (martha-johnston) [inline components/motor/gpio/motor_encoder.go:603]: Ideally I would like to have this or something so when you stop the motor the control loop doesn't just restart it since it's trying to maintain the current position, but as we saw @randhid it caused issues with the autotuning process so wondering if anyone has suggestions

File: components/motor/gpio/motor_encoder.go
Comment (martha-johnston) [inline components/motor/gpio/motor_encoder.go:190]: I kind of want to take the starting of the control loop out of the constructor so that the motor doesn't immediately turn on without a motor command, so that's why this `if` block would be necessary, but if we keep it in the constructor this can be removed

File: components/motor/gpio/loop.go
Comment (randhid) [inline components/motor/gpio/loop.go:12]: comment on why this is needed.

File: components/motor/gpio/motor_encoder.go
Comment (randhid) [inline components/motor/gpio/motor_encoder.go:453]: I'd rather refactor these into helper functions in `loop.go`

File: components/motor/gpio/motor_encoder.go
Comment (randhid) [inline components/motor/gpio/motor_encoder.go:426]: are we under a lock?

File: components/motor/gpio/motor_encoder.go
Comment (randhid) [reply components/motor/gpio/motor_encoder.go:593]: we could set the constant value to ignore the control loop, a speed of zero - would you like to test that out?

File: components/motor/gpio/motor_encoder.go
Comment (randhid) [inline components/motor/gpio/motor_encoder.go:463]: also a comment here that this is both velocity and position control

File: control/control_block.go
Comment (randhid) [inline control/control_block.go:18]: A comment that these are the way we're exposing command inputs into the control config. We're going to have to think a little about exposing functions so users don't access them as freely when they're coding their own control blocks. Can we use the controlBlockType  iota instead in your code?

**Review by randhid — CHANGES_REQUESTED** [review]
First pass done! Good job! Let's discuss the suggestions for stop - can we input a zero velocity command and try that to stop the motor as you want? Also - probably good to have Nico look at this too!

**Comment by martha-johnston** [comment]
@randhid just fixed the GoTo bug! will go through your comments now

File: components/motor/gpio/loop.go
Comment (martha-johnston) [reply components/motor/gpio/loop.go:12]: adding a comment about how it's to scale the power input over the 0 to +/- 255 range from the control config, but is it valid to assume that this 255 is correct? in the docs we just say to use +/- 255 as the max/min, but a user definitely could change that

Comment (martha-johnston) [reply components/motor/gpio/loop.go:12]: although this whole project has kind of been under the assumption they are using the bucket rover so maybe it's fine

File: components/motor/gpio/motor_encoder.go
Comment (martha-johnston) [reply components/motor/gpio/motor_encoder.go:426]: yes

File: control/control_block.go
Comment (martha-johnston) [reply control/control_block.go:18]: not sure how to create the new blockConfig without exposing this type but I'll keep working on this

File: components/motor/gpio/motor_encoder.go
Comment (martha-johnston) [inline components/motor/gpio/motor_encoder.go:113]: added a tuning variable here

File: components/motor/gpio/motor_encoder.go
Comment (randhid) [inline components/motor/gpio/motor_encoder.go:114]: does this freak out if they aren't set for a non-control loop motor?

File: components/motor/gpio/loop.go
Comment (randhid) [inline components/motor/gpio/loop.go:35]: Since these are hardcoded block config names, should we write a validator for these names in this file?

File: components/motor/gpio/loop.go
Comment (randhid) [reply components/motor/gpio/loop.go:52]: maybe also a more function-exposing name

File: components/motor/gpio/motor_encoder.go
Comment (randhid) [inline components/motor/gpio/motor_encoder.go:108]: Have you observed the loop canceling the context if misconfigured in the config or given bad inputs such as `NaN` input speeds?

**Review by randhid — CHANGES_REQUESTED** [review]
Looking way better, nice bug finding!

File: components/motor/gpio/loop.go
Comment (martha-johnston) [reply components/motor/gpio/loop.go:24]: no it does not :)

File: components/motor/gpio/motor_encoder.go
Comment (martha-johnston) [reply components/motor/gpio/motor_encoder.go:108]: I have not. when I didn't have the config set up correctly it just didn't work. I added PID to the validate function I am working on to make sure that's also created correctly

File: components/motor/gpio/motor_encoder.go
Comment (martha-johnston) [reply components/motor/gpio/motor_encoder.go:114]: no because it won't get to this line if it's a non-controlled motor

Comment (martha-johnston) [reply components/motor/gpio/motor_encoder.go:114]: currently only have the control loop set up for the right motor on bucket and left still works normally with rpm monitor

Comment (randhid) [reply components/motor/gpio/motor_encoder.go]: excellent test rig.

File: control/control_block.go
Comment (randhid) [inline control/control_block.go:24]: beautiful, thank you

File: components/motor/gpio/motor_encoder_test.go
Comment (randhid) [inline components/motor/gpio/motor_encoder_test.go:174]: beautiful test, but I'd like to move away from relying on the exported functions of encoders at least. If not a bother to add it as a dependency through a config and a setup function, please do so!

**Review by randhid — APPROVED** [review]
Beautiful work! Let's double-check physical behaviour on Monday, no hangs and a go test with -race and a bunch of counts in a canon shell on that added function.

Otherwise - nits.

**Comment by martha-johnston** [comment]
> Beautiful work! Let's double-check physical behaviour on Monday, no hangs and a go test with -race and a bunch of counts in a canon shell on that added function.
> 
> Otherwise - nits.

Sounds good, and will wait for a review from Nicolas when he's back

File: control/trapezoid_velocity_profile.go
Comment (martha-johnston) [inline control/trapezoid_velocity_profile.go:125]: @randhid thoughts on this? on line 61 of this file I had previously added an additional condition `|| setPoint == 0` to the if statement because when the set point was 0, the values in this block weren't being set because `setPoint` and `lastsetPoint` both equaled 0. testing the behavior just now, I noticed that on a goTo 0 call, the motor moved extremely slow, which I realized was because this if block was always true when setPoint == 0, but the values in the block should only be calculated at the beginning of a new trap profile. by setting lastsetPoint to NaN, the original if statement behaves as necessary, but I'm not sure if we don't want to set a number to NaN, that seems hacky

Comment (randhid) [reply control/trapezoid_velocity_profile.go:125]: We can either do the following:

- make this a behaviour in `State` and `SetState` in the motor - where we ignore the set point if it it being set in a specific way by GoTo - which is the case you're hitting.
- merge this in and test further, I like the fact that the loop is capable of handling it - but we will have to figure out how that works with the SetVelocity loop in sensor-controlled base.

File: components/motor/gpio/loop.go
Comment (npmenard) [inline components/motor/gpio/loop.go:23]: nit: scale power to +/- 1 from PID output

File: components/motor/gpio/loop.go
Comment (npmenard) [inline components/motor/gpio/loop.go:13]: nit: can use a gain block to scale down the power output

File: components/motor/gpio/loop.go
Comment (npmenard) [inline components/motor/gpio/loop.go:73]: nit: `if err := cLoop.Start(); err != nil {return err}`

File: components/motor/gpio/motor_encoder.go
Comment (npmenard) [inline components/motor/gpio/motor_encoder.go:185]: are we re-creating a new control loop each time rpmMonitorStart is called? or is it a one of?

File: control/control_loop.go
Comment (npmenard) [inline control/control_loop.go:288]: is there a way we could have this logic in the PID block?

**Review by npmenard — APPROVED** [review]
LGTM just a couple of comments

File: components/motor/gpio/loop.go
Comment (martha-johnston) [reply components/motor/gpio/loop.go:23]: removing comment and adding a gain block

File: components/motor/gpio/motor_encoder.go
Comment (martha-johnston) [reply components/motor/gpio/motor_encoder.go:185]: yes we are recreating with each every movement call so that the loop can be stopped when the position is reached instead of the motor holding position. I mostly implemented it this way so while I was testing with app it would be less annoying so I'm not opposed to changing it to be constantly running until the motor is closed

File: control/trapezoid_velocity_profile.go
Comment (randhid) [reply control/trapezoid_velocity_profile.go:125]: @npmenard do you prefer the `math.Nan` approach or having the driver implementing the control package handle this case?

File: control/control_loop.go
Comment (martha-johnston) [reply control/control_loop.go:288]: like add it to the PID config? I can for sure

Comment (martha-johnston) [reply control/control_loop.go:288]: since we will be discussing changes to the control config in a couple weeks, I am thinking I'll leave this for now

**Comment by martha-johnston** [comment]
Just made a couple more changes to how the control loop is created and stopped based on the motor commands. I think there is a better way to do this with some changes to the control package, but for now this is as clean as I can get the setup without there being major racing issues

File: components/motor/gpio/motor_encoder.go
Comment (martha-johnston) [inline components/motor/gpio/motor_encoder.go:584]: specifically, setting the loop to nil fixed the issue I was having that the control loop wasn't being closed as expected and multiple control loops were conflicting with each other. is it fine to set the loop to nil in this way or is there a better way to go about this?

Comment (randhid) [reply components/motor/gpio/motor_encoder.go:586]: to answer your question, this is fine, it is conceptually the same pattern as calling a cancel function on a context, which will make a context nil I believe.

Comment (randhid) [reply components/motor/gpio/motor_encoder.go:586]: but we should be setting that channel to nil in the control loop if needed.

**Pull request merged**
